Handle empty index case in LuceneSyntheticSourceChangesSnapshot

martijnvg opened the pull request:
In case of synthetic recovery source when the mapping is empty.

A test that reproduces failure in #118955 consistently with a potential fix.

`MapperService#updateMapping(...)` doesn't set the mapper field if a mapping has no fields, which is what is used in InternalEngine#newChangesSnapshot(...) . This happens when `newMappingMetadata` variable in `MapperService updateMapping(...)` is `null`. Causing an assertion to trip. This change adjusts that assertion to handle an empty index.

Closes #118955

martijnvg committed:
Return empty snapshot if index is empty.

In case of synthetic recovery source the mapping is empty.

Closes #118955

martijnvg committed:
fix mute file

martijnvg committed:
iter

jimczi commented on server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -3159,17 +3159,21 @@ public Translog.Snapshot newChangesSnapshot(
-                    accessStats,
-                    config().getIndexSettings().getIndexVersionCreated()
-                );
+                if (searcher.getIndexReader().maxDoc() == 0) {
```
Maybe add a small comment to explain why we're doing that?

jimczi reviewed (approved):
LGTM

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

martijnvg committed:
added comment

martijnvg committed:
just alter assertion instead of returning an empty snapshot.
(some tests failed, which expected an exception, which no longer occurs when returning an empty snapshot)

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_recovery_source_empty_shard

martijnvg committed:
use index api

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_recovery_source_empty_shard

martijnvg committed:
iter

martijnvg commented:
@elasticmachine run elasticsearch-ci/part-1

martijnvg committed:
Merge remote-tracking branch 'es/main' into synthetic_recovery_source_empty_shard

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 118996`
